File: HyungJoonLEE/2022-07-25-Remove-Nth-Node-From-End-of-List.c
#include "common.h"

struct ListNode {
    int val;
    struct ListNode *next;
};


struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode* list = head;
    int count = 0, chaser = 0;

    while(list) {
        count++;
        list = list->next;
    }

    if (count == 1 && n == 1) return NULL;

    count = count - n;
    list = head;
    if (chaser == count) return head = list->next;

    while(chaser != count) {
        chaser++;
        if (chaser == count) list->next = list->next->next;
        list = list->next;
    }
    return head;
}

File: Hyerin Cheon/2022-07-25-Remove-Nth-Node-From-End-Of-List.js
function remoceNthFromEnd(head, n){
  let dummyHead = new ListNode(-Infinity);
  dummyHead.next = head;

  // need to reference for the output since the dummyHead is increment all the way
  let resultHead = dummyHead;

  let tail = head;
  let count = 0;

  // move tail by n
  while(count < n){
    tail = tail.next;
    count++;
  }

  let removeNode = head;
  let prev = dummyHead;

  while(tail){
    tail = tail.next;
    removeNode = removeNode.next;
    prev = prev.next;
  }
  // if tail is null
  // break the pointer and link to new node
  prev.next = removeNode.next;

  return resultHead.next;
}